x-pack/plugins/encrypted_saved_objects/server/lib/encrypted_saved_objects_service.ts

```diff
@@ -0,0 +1,289 @@
+      if (attributeValue == null) {
+        continue;
+      }
+
+      if (typeof attributeValue !== 'string') {
+        this.audit.decryptAttributeFailure(attributeName, type, id);
+        throw new Error(
+          `Encrypted "${attributeName}" attribute should be a string, but found ${typeDetect(
+            attributeValue
+          )}`
+        );
+      }
+
+      try {
+        decryptedAttributes[attributeName] = await this.crypto.decrypt(
+          attributeValue,
+          encryptionAAD
+        );
+      } catch (err) {
+        this.log.error(`Failed to decrypt "${attributeName}" attribute: ${err.message || err}`);
+        this.audit.decryptAttributeFailure(attributeName, type, id);
+
+        throw new EncryptionError(
+          `Unable to decrypt attribute "${attributeName}"`,
+          attributeName,
+          err
+        );
+      }
+    }
+
+    // Normally we expect all registered to-be-encrypted attributes to be defined, but if it's
+    // not the case we should collect and log them to make troubleshooting easier.
+    const decryptedAttributesKeys = new Set(Object.keys(decryptedAttributes));
+    if (decryptedAttributesKeys.size !== typeRegistration.attributesToEncrypt.size) {
+      this.log.debug(
+        `The following attributes of saved object "${type}:${id}" should have been decrypted: ${Array.from(
+          typeRegistration.attributesToEncrypt
+        )}, but found only: ${Array.from(decryptedAttributesKeys)}`
+      );
+    }
+
+    if (decryptedAttributesKeys.size === 0) {
+      return attributes;
+    }
+
+    this.audit.decryptAttributesSuccess(decryptedAttributesKeys, type, id);
+
+    return {
+      ...attributes,
+      ...decryptedAttributes,
+    };
+  }
+
+  /**
+   * Generates string representation of the Additional Authenticated Data based on the specified saved
+   * object type and attributes.
+   * @param typeRegistration Saved object type registration parameters.
+   * @param id Id of the saved object.
+   * @param attributes All attributes of the saved object instance of the specified type.
+   */
+  private getAAD<T extends Record<string, unknown>>(
+    typeRegistration: EncryptedSavedObjectTypeRegistration,
+    id: string,
+    attributes: T
+  ) {
+    // Collect all attributes (both keys and values) that should contribute to AAD.
+    const attributesAAD = Array.from(Object.entries(attributes)).filter(([attributeKey]) => {
+      return (
+        !typeRegistration.attributesToEncrypt.has(attributeKey) &&
+        (typeRegistration.attributesToExcludeFromAAD == null ||
+          !typeRegistration.attributesToExcludeFromAAD.has(attributeKey))
+      );
+    });
+
+    if (attributesAAD.length === 0) {
+      this.log.debug(
+        `The AAD for saved object "${typeRegistration.type}:${id}" does not include any attributes.`
+      );
+    }
+
+    // TODO: Use JSON.stringify instead of template string? Slower, but respects value types, e.g.
```

Comment:
__question:__ what do you think about this comment? Calling `toString` is faster, but at the same time I don't think `JSON.stringify` will be a bottleneck.

Reply:
Good point! Will definitely switch to `json-stable-stringify`.